File: backend_python_legacy/app/api/v1/presets.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException, status

from app.models.schemas import ErrorResponse, PresetSummary, Scenario

router = APIRouter(tags=["presets"])


def find_data_dir() -> Path:
    """Поиск директории эталонных сценариев 'data' или 'Данные'."""
    names = ["data", "Данные"]
    candidates = []
    for name in names:
        candidates.extend(
            [
                Path.cwd() / name,
                Path.cwd().parent / name,
                Path(__file__).resolve().parents[4] / name,
                Path(__file__).resolve().parents[3] / name,
                Path(__file__).resolve().parents[2] / name,
            ]
        )
    for c in candidates:
        if c.exists() and c.is_dir():
            return c
    return Path("data")


@router.get(
    "/presets",
    response_model=list[PresetSummary],
    summary="Получение списка доступных пресетов сценариев",
)
def get_presets() -> list[PresetSummary]:
    """Возвращает метаданные и параметры доступных файлов сценариев."""
    data_dir = find_data_dir()
    if not data_dir.exists():
        return []

    summaries: list[PresetSummary] = []
    for p in sorted(data_dir.glob("*.json")):
        try:
            content = json.loads(p.read_text(encoding="utf-8"))
            if content.get("schema_version") != "cosmo-A-1.0":
                continue
            meta = content.get("meta", {})
            env = content.get("environment", {})
            des = content.get("design", {})
            ground = content.get("ground_sites", [])

            client_count = sum(1 for g in ground if g.get("role") == "client")
            gateway_count = sum(1 for g in ground if g.get("role") == "gateway")

            summaries.append(
                PresetSummary(
                    id=meta.get("id", p.stem),
                    title=meta.get("title", p.name),
                    filename=p.name,
                    satellite_count=len(des.get("satellites", [])),
                    planes_count=len(des.get("planes", [])),
                    client_count=client_count,
                    gateway_count=gateway_count,
                    horizon_s=int(env.get("horizon_s", 86400)),
                    step_s=int(env.get("step_s", 120)),
                    launch_stage=int(des.get("launch_stage", 3)),
                    isl_range_km=float(env.get("isl_range_km", 3000.0)),
                )
            )
        except Exception:
            continue

    return summaries


@router.get(
    "/presets/{name}",
    response_model=Scenario | dict[str, Any],
    summary="Получение конфигурации пресета по имени файла",
    responses={
        200: {"description": "Полная спецификация сценария в формате cosmo-A-1.0"},
        404: {"model": ErrorResponse, "description": "Сценарий не найден в каталоге пресетов"},
        500: {"model": ErrorResponse, "description": "Ошибка чтения файла сценария"},
    },
)
def get_preset(name: str) -> dict[str, Any]:
    """Возвращает полное содержимое сценария в формате JSON."""
    data_dir = find_data_dir()
    filename = name if name.endswith(".json") else f"{name}.json"
    preset_path = data_dir / filename

    if not preset_path.exists():
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Сценарий '{name}' не найден в каталоге пресетов.",
        )

    try:
        return json.loads(preset_path.read_text(encoding="utf-8"))
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ошибка чтения файла сценария: {e}",
        )

File: backend/crates/cosmo-api/src/handlers/presets.rs
use axum::{extract::Path, http::StatusCode, Json};
use cosmo_core::Scenario;
use serde_json::{json, Value};
use std::{fs, path::Path as StdPath};

use crate::config::find_data_dir;

/// Получение списка всех доступных конфигураций пресетов.
pub async fn get_presets_endpoint() -> Json<Value> {
    let data_dir = find_data_dir();
    let mut summaries = Vec::new();

    if let Ok(entries) = fs::read_dir(&data_dir) {
        let mut paths: Vec<_> = entries
            .filter_map(|e| e.ok())
            .map(|e| e.path())
            .filter(|p| p.extension().map_or(false, |ext| ext == "json"))
            .collect();
        paths.sort();

        for p in paths {
            if let Ok(content_str) = fs::read_to_string(&p) {
                if let Ok(scenario) = serde_json::from_str::<Scenario>(&content_str) {
                    if scenario.schema_version.as_deref() != Some("cosmo-A-1.0") {
                        continue;
                    }
                    let client_count = scenario
                        .ground_sites
                        .iter()
                        .filter(|g| g.role == "client")
                        .count();
                    let gateway_count = scenario
                        .ground_sites
                        .iter()
                        .filter(|g| g.role == "gateway")
                        .count();
                    let filename = p
                        .file_name()
                        .unwrap_or_default()
                        .to_string_lossy()
                        .to_string();
                    let stem = p
                        .file_stem()
                        .unwrap_or_default()
                        .to_string_lossy()
                        .to_string();

                    let meta_id = scenario
                        .meta
                        .as_ref()
                        .and_then(|m| m.get("id"))
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string())
                        .unwrap_or(stem);

                    let meta_title = scenario
                        .meta
                        .as_ref()
                        .and_then(|m| m.get("title"))
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string())
                        .unwrap_or_else(|| filename.clone());

                    summaries.push(json!({
                        "id": meta_id,
                        "title": meta_title,
                        "filename": filename,
                        "satellite_count": scenario.design.satellites.len(),
                        "planes_count": scenario.design.planes.len(),
                        "client_count": client_count,
                        "gateway_count": gateway_count,
                        "horizon_s": scenario.environment.horizon_s,
                        "step_s": scenario.environment.step_s,
                        "launch_stage": scenario.design.launch_stage,
                        "isl_range_km": scenario.environment.isl_range_km
                    }));
                }
            }
        }
    }

    Json(json!(summaries))
}

/// Получение содержимого сценария пресета по имени.
pub async fn get_preset_endpoint(
    Path(name): Path<String>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    let data_dir = find_data_dir();
    let file_stem = StdPath::new(&name)
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or(&name);
    let filename = if file_stem.ends_with(".json") {
        file_stem.to_string()
    } else {
        format!("{}.json", file_stem)
    };
    let preset_path = data_dir.join(&filename);

    let canonical_data = data_dir.canonicalize().unwrap_or_else(|_| data_dir.clone());
    let is_safe = preset_path
        .canonicalize()
        .map(|p| p.starts_with(&canonical_data))
        .unwrap_or(false);

    if !preset_path.exists() || !is_safe {
        return Err((
            StatusCode::NOT_FOUND,
            Json(json!({
                "message": format!("Сценарий '{}' не найден в каталоге пресетов.", name)
            })),
        ));
    }

    match fs::read_to_string(&preset_path) {
        Ok(content) => match serde_json::from_str::<Value>(&content) {
            Ok(json_val) => Ok(Json(json_val)),
            Err(e) => Err((
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({
                    "message": format!("Ошибка парсинга сценария: {}", e)
                })),
            )),
        },
        Err(e) => Err((
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({
                "message": format!("Ошибка чтения файла сценария: {}", e)
            })),
        )),
    }
}
